banderwagon/element.go

```diff
@@ -72,36 +78,54 @@ func (p Element) BytesUncompressed() [UncompressedSize]byte {
 func BatchNormalize(elements []*Element) {
 	// The elements slice might contain duplicate pointers,
 	// dedupe them to avoid double work.
-	dedupedElements := make([]*Element, 0, len(elements))
+	mapDedupedElements := make(map[*Element]struct{}, len(elements))
 	for _, e := range elements {
-		found := false
-		for i := range dedupedElements {
-			if dedupedElements[i] == e {
-				found = true
-				break
-			}
-		}
-		if found {
-			continue
-		}
+		mapDedupedElements[e] = struct{}{}
+	}
+	dedupedElements := make([]*Element, 0, len(mapDedupedElements))
+	for e := range mapDedupedElements {
 		dedupedElements = append(dedupedElements, e)
 	}
 
-	// Collect all z co-ordinates
-	zs := make([]fp.Element, len(dedupedElements))
+	invs := make([]fp.Element, len(elements))
+	zeroes := make([]bool, len(elements))
+	accumulator := fp.One()
+
+	// batch invert all points[].Z coordinates with Montgomery batch inversion trick
+	// (stores points[].Z^-1 in result[i].X to avoid allocating a slice of fr.Elements)
 	for i := 0; i < len(dedupedElements); i++ {
-		zs[i] = dedupedElements[i].inner.Z
+		if dedupedElements[i].inner.Z.IsZero() {
+			zeroes[i] = true
+			continue
+		}
+		invs[i] = accumulator
+		accumulator.Mul(&accumulator, &dedupedElements[i].inner.Z)
 	}
 
-	// Invert z co-ordinates
-	zInvs := fp.BatchInvert(zs)
+	var accInverse fp.Element
+	accInverse.Inverse(&accumulator)
 
-	// Multiply x and y by zInv
-	for i, e := range dedupedElements {
-		e.inner.X.Mul(&e.inner.X, &zInvs[i])
-		e.inner.Y.Mul(&e.inner.Y, &zInvs[i])
-		e.inner.Z.SetOne()
+	for i := len(dedupedElements) - 1; i >= 0; i-- {
+		if zeroes[i] {
+			// do nothing, (X=0, Y=0) is infinity point in affine
+			continue
+		}
+		invs[i].Mul(&invs[i], &accInverse)
+		accInverse.Mul(&accInverse, &dedupedElements[i].inner.Z)
 	}
+
+	// batch convert to affine.
+	parallel.Execute(len(dedupedElements), func(start, end int) {
+		for i := start; i < end; i++ {
+			if zeroes[i] {
+				// do nothing, (X=0, Y=0) is infinity point in affine
+				continue
+			}
+			dedupedElements[i].inner.X.Mul(&dedupedElements[i].inner.X, &invs[i])
+			dedupedElements[i].inner.Y.Mul(&dedupedElements[i].inner.Y, &invs[i])
+			dedupedElements[i].inner.Z = fp.One()
+		}
+	})
```

Comment:
Nothing entirely interesting here, it's a very similar code we have in some other place to parallelize normalizing elements. 

Unfortunately, I couldn't reuse other similar implementations since they worked in `bandersnatch.Element`, and here I need `banderwagon.Element` normalization.

I could re-use the other implementation if I'm OK with doing some "point copying" but that's quite unfortunate. (i.e: having to "copy points" just because it's hard to make a generic implementation of this code in Go).